mayastor/src/bdev/dev/nvmx/channel.rs

```diff
@@ -63,24 +65,75 @@ pub struct NvmeIoChannelInner<'a> {
     pub qpair: *mut spdk_nvme_qpair,
     poll_group: NonNull<spdk_nvme_poll_group>,
     poller: poller::Poller<'a>,
+    // Flag to indicate the shutdown state of the channel.
+    // We need such a flag to differentiate between channel reset and shutdown.
+    // Channel reset is a reversible operation, which is followed by
+    // reinitialize(), which 'resurrects' channel (i.e. recreates all its
+    // I/O resources): such behaviour is observed during controller reset.
+    // Shutdown, in contrary, means 'one-way' ticket for the channel, which
+    // doesn't assume any further resurrections: such behaviour is seen
+    // upon controller shutdown. Being able to differentiate between these
+    // 2 states allows controller reset to behave properly in parallel with
+    // shutdown (if case reset is initiated before shutdown), and
+    // not to reinitialize channels already processed by shutdown logic.
+    is_shutdown: bool,
 }
 
 impl NvmeIoChannelInner<'_> {
-    /// Resets channel, making it unusable till reinitialize() is called.
+    /// Reset channel, making it unusable till reinitialize() is called.
     pub fn reset(&mut self) -> i32 {
         if self.qpair.is_null() {
             return 0;
         }
 
-        0
+        self._abort_queue();
+
+        let q = self.qpair;
+        self.qpair = std::ptr::null_mut();
+
+        unsafe {
+            spdk_nvme_ctrlr_disconnect_io_qpair(q);
+            spdk_nvme_ctrlr_free_io_qpair(q)
+        }
+    }
+
+    /// Checks whether the I/O channel is shutdown.
+    pub fn is_shutdown(&self) -> bool {
+        self.is_shutdown
+    }
+
+    /// Shutdown I/O channel and make it completely unusable for I/O.
+    pub fn shutdown(&mut self) -> i32 {
+        if self.is_shutdown {
+            return 0;
+        }
+
+        let rc = self.reset();
+        if rc == 0 {
+            self.is_shutdown = true;
+        }
+        rc
     }
 
-    /// Reinitializes channel after reset.
+    /// Aborts all the requests in the qpair.
+    fn _abort_queue(&self) {
```

Comment:
we dont need to have _ prefixed I think, the fn already is private.

Reply:
This is touched in https://github.com/openebs/Mayastor/pull/648#discussion_r562084675, it's to signify callbacks!